components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/OAuthAuthenticationInterceptor.java

```diff
@@ -42,90 +47,55 @@
             return;
         }
-
-        inMessage.put(RestApiConstants.REQUEST_AUTHENTICATION_SCHEME, RestApiConstants.OAUTH2_AUTHENTICATION);
-
-        if(handleRequest(inMessage, null)){
-            /*String requestedTenant = ((ArrayList) ((TreeMap) (inMessage.get(Message.PROTOCOL_HEADERS))).get("X-WSO2_Tenant")).get(0).toString();
-            if(requestedTenant!=null){
-                RestApiUtil.setThreadLocalRequestedTenant(requestedTenant);
-            }
-            else {
-                RestApiUtil.unsetThreadLocalRequestedTenant();
-            }*/
-            if(logger.isDebugEnabled()) {
-                logger.debug("User logged into Web app using OAuth Authentication");
-            }
-        }
-        else{
-            throw new AuthenticationException("Unauthenticated request");
+        //identify Oauth2 and JWT tokens
+        if (accessToken.contains(RestApiConstants.DOT)) {
+            inMessage.put(RestApiConstants.REQUEST_AUTHENTICATION_SCHEME, RestApiConstants.JWT_AUTHENTICATION);
+        } else {
+            inMessage.put(RestApiConstants.REQUEST_AUTHENTICATION_SCHEME, RestApiConstants.OPAQUE_AUTHENTICATION);
         }
-    }
 
-    /**
-     * This method will initialize Web APP authenticator to validate incoming requests
-     * Here we will get implementation class and create object of it.
-     */
-    public void initializeAuthenticator() throws APIManagementException {
         try {
-            //TODO Retrieve this class name from configuration and let it configurable.
-            //  authenticator = (WebAppAuthenticator) APIUtil.getClassForName(
-            //      RestApiConstants.REST_API_WEB_APP_AUTHENTICATOR_IMPL_CLASS_NAME).newInstance();
-            authenticator = new WebAppAuthenticatorImpl();
-        } catch (Exception e) {
-            throw new APIManagementException("Error while initializing authenticator of " + "type: ",e);
-        }
-
-    }
-
-    /**
-     * authenticate requests received at the REST API endpoint, using HTTP OAuth headers as the authentication
-     * mechanism. This method returns a null value which indicates that the request to be processed.
-     */
-    public boolean handleRequest(Message message, ClassResourceInfo resourceInfo) {
-
-        if (authenticator == null) {
-            try {
-                initializeAuthenticator();
-            } catch (APIManagementException e) {
-                if (logger.isDebugEnabled()) {
-                    logger.debug(" Initializing the authenticator resulted in an exception", e);
-                }else{
-                    logger.error(e.getMessage());
-                }
-                return false;
+            if (logger.isDebugEnabled()) {
+                logger.debug(String.format("Authenticating request: " + inMessage.getId()));
             }
-          }
-            try {
+
+            if (authenticatorMap.get(inMessage.get(RestApiConstants.REQUEST_AUTHENTICATION_SCHEME)).authenticate(inMessage)) {
```

Comment:
Let's break into two steps : 1) getting the authenticator -> 2) do authenticate
Also add a debug log with the selected authenticator along with the masked access token.